arangod/GeneralServer/GeneralServer.cpp

```diff
@@ -120,19 +120,37 @@ void GeneralServer::startListening() {
   }
 }
 
+/// stop accepting new connections
 void GeneralServer::stopListening() {
   for (std::unique_ptr<Acceptor>& acceptor : _acceptors) {
     acceptor->close();
   }
+}
 
+/// stop connections
+void GeneralServer::stopConnections() {
   // close connections of all socket tasks so the tasks will
   // eventually shut themselves down
   std::lock_guard<std::recursive_mutex> guard(_tasksLock);
-  _commTasks.clear();
+  for (auto const& pair : _commTasks) {
+    pair.second->stop();
+  }
 }
 
 void GeneralServer::stopWorking() {
   _acceptors.clear();
+  std::unique_lock<std::recursive_mutex> guard(_tasksLock);
+  auto now = std::chrono::system_clock::now();
+  while (!_commTasks.empty()) {  // CommTasks should deregister themselves
+    guard.unlock();
+    std::this_thread::yield();
+    if ((std::chrono::system_clock::now() - now) < std::chrono::seconds(3)) {
```

Comment:
Shouldn't we check for if *more* than 3 seconds passed, not less?
```suggestion
    if ((std::chrono::system_clock::now() - now) > std::chrono::seconds(3)) {
```